invoices/invoice_expiry_watcher.go

```diff
@@ -0,0 +1,215 @@
+func (ew *InvoiceExpiryWatcher) nextExpiry() time.Duration {
+	if !ew.expiryQueue.Empty() {
+		top := ew.expiryQueue.Top().(*invoiceExpiry)
+		return top.Expiry.Sub(ew.clock.Now())
+	}
+
+	return time.Duration(math.MaxInt64)
+}
+
+// cancelExpiredInvoices will cancel all expired invoices and removes them from
+// the expiry queue.
+func (ew *InvoiceExpiryWatcher) cancelExpiredInvoices() {
+	for {
+		if ew.expiryQueue.Empty() {
+			break
+		}
+
+		top := ew.expiryQueue.Top().(*invoiceExpiry)
+		if time.Since(top.Expiry) > 0 {
+			log.Infof("Cancelling expired invoice '%v', expiry: %v",
+				top.PaymentHash, top.Expiry)
+
+			if err := ew.registry.CancelInvoice(top.PaymentHash); err != nil {
+				log.Errorf("Unable to cancel invoice: %v", top.PaymentHash)
+			}
+			ew.expiryQueue.Pop()
+		}
+	}
+}
+
+// mainLoop receives invoice events and handles cancellaton. Only new
+// invoices that are not already settled or canceled will be added to the queue.
+func (ew *InvoiceExpiryWatcher) mainLoop() {
+	defer ew.wg.Done()
+
+	for {
+		// Cancel any invoices that may have expired.
+		ew.cancelExpiredInvoices()
+
+		select {
+		case <-time.After(ew.nextExpiry()):
+			// Wait until the next invoice expires, then cancel expired invoices.
+			continue
+
+		case invoice := <-ew.subscription.NewInvoices:
+			// Only care about invoices that are not canceled or settled already.
+			if invoice.State != channeldb.ContractCanceled &&
+				invoice.State != channeldb.ContractSettled {
+				// Try to add the invoice to the expiry watcher queue.
+				if err := ew.addInvoice(invoice); err != nil {
+					log.Errorf("Error adding invoice to the expiry watcher queue: %v", err)
+				}
+			}
+
+		case <-ew.subscription.SettledInvoices:
+			// Do nothing with settled invoices. The expiry watcher will try
+			// to cancel (which will be ignored).
+
+		case <-ew.quit:
+			return
+		}
+	}
+}
+
+// decodePaymentHash is a helper function to decode the payment hash from the
+// payment request of the passed Invoice. Returns a zero hash if decoding fails.
+// TODO: Remove this function, once InvoiceSubscription forwards payment hashes
+// for all invoices or when channeldb.Invoice holds the payment hash.
+func (ew *InvoiceExpiryWatcher) decodePaymentHash(
+	invoice *channeldb.Invoice) (lntypes.Hash, error) {
+
+	decodedInvoice, err := zpay32.Decode(
+		string(invoice.PaymentRequest), ew.netParams)
+
+	if err != nil {
+		return lntypes.ZeroHash, err
+	}
+
+	return *decodedInvoice.PaymentHash, nil
+}
+
+// addInvoice acquires the payment hash for the invoice and then passes it to
+// the expiry watcher queue.
+func (ew *InvoiceExpiryWatcher) addInvoice(invoice *channeldb.Invoice) error {
+	// The payment hash is decoded from the payment request that is stored
+	// in the invoice.
+	paymentHash, err := ew.decodePaymentHash(invoice)
+	if err != nil {
+		return fmt.Errorf("Could not decode payment hash for invoice: %v", invoice)
```

Comment:
Again this code should not be here, it's an outdated version and you only see this because of a rebase error on my side. I'll try to fix the commit structure again.